Implement the quiz pagination frontend

m1r0 opened the pull request:
Fixes #4480 

### Changes proposed in this Pull Request

This PR adds the quiz pagination frontend UI and refactors some of the default quiz action buttons code. It does not implement the session needed for the quiz form submission to work properly, so no need to test that.

### Testing instructions

* Enable the quiz pagination feature:
```php
add_filter( 'sensei_feature_flag_quiz_pagination', '__return_true' );
```
* Make a new course with a lesson.
* Make a quiz with multiple questions (>=8).
* Enable the pagination by updating the quiz pagination settings (see #4387).
* Enroll in the course.
* Open the quiz in the frontend.
* Don't use the save/reset/complete buttons. They will be fixed when the quiz session is implemented (#4392).
* Make sure the question listing is paginated.
* Make sure the numeric pagination and the next/prev buttons work.
* Check if the question numbers are correct.
* Edit the quiz and enable the `Allow Retakes` setting and make sure the `Reset` button is present in the frontend.
* Make sure you see the `Complete` button on the last page.
* Make sure you have at least 8 pages and go to page 4 to see the full width of the numeric pagination.
* Make sure the pagination looks good on the more popular WP themes.
* Make sure the pagination looks good on mobile.
* Make sure the question listing is paginated even if not enrolled in the course.
* Make the quiz pagination per page number bigger than the questions count and make sure the pagination UI is not loading. You should see the default action buttons instead.
* Disable the quiz pagination and make sure the default quiz action buttons work as expected.

### New Hooks

* `sensei_quiz_pagination_args` - Filters the quiz questions paginate links args.
* `sensei_the_question_number` - Filters the number of the current question within the quiz question loop.

### Changed Templates

* `single-quiz.php` - Explicitly provide the question number.

### Screenshots

Desktop:
<img width="662" alt="Screenshot on 2021-12-14 at 15-51-34" src="https://user-images.githubusercontent.com/1612178/146012162-f395e4b4-6905-459d-b75a-15996d744036.png">
<img width="851" alt="Screenshot on 2021-12-14 at 15-54-04" src="https://user-images.githubusercontent.com/1612178/146012191-d13eb774-096f-4bf8-8f11-51869ce8487a.png">


Mobile:
<img width="353" alt="Screenshot on 2021-12-14 at 15-55-39" src="https://user-images.githubusercontent.com/1612178/146012217-562b23d1-bfc7-4c00-a019-4aab2799706b.png">
<img width="341" alt="Screenshot on 2021-12-14 at 15-54-47" src="https://user-images.githubusercontent.com/1612178/146012235-c9cf39d4-2402-43ef-a2b5-2f4e5cecc42a.png">

m1r0 commented on includes/class-sensei-quiz.php:
```diff
@@ -1231,6 +1231,62 @@ public function quiz_has_no_questions() {
+	 *
+	 * @return bool
+	 */
+	public function can_take_quiz( int $quiz_id = null, int $user_id = null ): bool {
```
This method uses the same logic used in `action_buttons()`.

donnapep replied:
Does this function need to be `public`?

m1r0 commented on includes/class-sensei-quiz.php:
```diff
@@ -1285,6 +1341,7 @@ public static function start_quiz_questions_loop() {
 		$sensei_question_loop['questions']      = [];
 		$sensei_question_loop['posts_per_page'] = -1;
 		$sensei_question_loop['current_page']   = 1;
+		$sensei_question_loop['total_pages']    = 1;
```
This will make pagination easier. Woo is doing something [similar](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/loop/pagination.php#L22).

m1r0 commented on templates/single-quiz/pagination.php:
```diff
@@ -0,0 +1,97 @@
+						href="<?php echo esc_attr( add_query_arg( 'quiz-page', $sensei_question_loop['current_page'] - 1 ) ); ?>"
+						class="wp-block-button__link button quiz-submit complete sensei-stop-double-submission"
+					>
+						<?php esc_attr_e( '&lt; Prev', 'sensei-lms' ); ?>
```
Had to replace `Previous` with `Prev` to save some space. To understand the reasoning, check out the screenshots.

m1r0 replied:
Maybe I should stick with the `Previous` label and show the buttons on a new line?

donnapep replied:
I think this may be a concern for translators. People probably won't know what "Prev" means in order to translate it. Another point related to translations is that "Previous" might translate to a longer word in another language, so it would be nice if we could get this to look good with longer strings. If both buttons don't always fit on one row then that seems fine.

m1r0 replied:
Makes sense!

> It would be nice if we could get this to look good with longer strings

Here's an easy to do suggestion:

<img width="678" alt="Screenshot on 2021-12-13 at 15-06-20" src="https://user-images.githubusercontent.com/1612178/145818785-f4422dd2-8608-44c4-ab0f-52b2e351a63a.png">

Let me know if this could work or we should explore other options.

merkushin commented on includes/class-sensei-quiz.php:
```diff
@@ -1231,6 +1231,62 @@ public function quiz_has_no_questions() {
+		}
+
+		$lesson_status_comment_id = is_array( $lesson_status )
+			? $lesson_status[0]->comment_ID
```
I got these notices for a new quiz. 
When I tested it with a previously created quiz, I had no notices.

m1r0 replied:
Nice catch! It seems this happens when the lesson doesn't have a course initially. Fixed in https://github.com/Automattic/sensei/pull/4502/commits/f2a9fdb9424ad00d3b7dd98e61b9d4fb0632b42e.

merkushin reviewed (commented):
The pagination works well 👍 
However, I was confused with Reset/Save.
Could you tell me when the Reset link should appear?
And is it okay to click Save in the middle of the quiz?
Probably, I did something wrong.

m1r0 commented:
Hey, thanks for testing it! I've updated the PR description to make things clear.

> Could you tell me when the Reset link should appear?

You need to enable the `Allow Retakes` from the quiz settings. It's on the same screen as the pagination settings.

> And is it okay to click Save in the middle of the quiz?

The save/reset/complete buttons are not working at the moment. They will be fixed after the quiz session is implemented.

sruj09 commented:
Looks good @m1r0 🙌
One minor update, can we add the arrow icons in the buttons instead of the < and >?

![image](https://user-images.githubusercontent.com/72030923/145310363-c71ba402-3751-4143-9d6b-52a3d6309f97.png)

Here is the Figma link for the assets: https://www.figma.com/file/o2Xnqd2zgjaMaJFPmvNOx3/Sensei-Blocks?node-id=20010919%3A55852

donnapep commented:
Looks like there's an issue on Divi:

![Screen Shot 2021-12-09 at 8 15 58 AM](https://user-images.githubusercontent.com/1190420/145403339-1d90f2fe-0069-43b8-9ca7-f560a7980fc7.jpg)

donnapep commented on includes/class-sensei-quiz.php:
```diff
@@ -1405,75 +1461,49 @@ public static function the_user_status_message( $quiz_id ) {
+						<?php esc_attr_e( 'Reset Quiz', 'sensei-lms' ); ?>
+					</button>
+
+					<input type="hidden" name="woothemes_sensei_reset_quiz_nonce" id="woothemes_sensei_reset_quiz_nonce" value="<?php echo esc_attr( wp_create_nonce( 'woothemes_sensei_reset_quiz_nonce' ) ); ?>" />
```
Nice one on only rendering the nonces when needed.

donnapep commented on assets/css/frontend.scss:
```diff
@@ -370,6 +370,88 @@ a.sensei-certificate-link {
+	gap: 1.5em;
+	margin: 1.618em 0;
+
+	@media (max-width: 667px) {
```
Could we use a mobile first approach here and add base styles for mobile, then add more styles on top of that for wider viewports (i.e. switch from `max-width` to `min-width` media queries)?

m1r0 replied:
Good suggestion. Using mobile first should be less complicated in general. Fixed in 8ba389fc02e281504f468552b704fe16bf441da6.

donnapep commented on templates/single-quiz/pagination.php:
```diff
@@ -0,0 +1,101 @@
+	<div class="sensei-quiz-pagination__actions">
+		<?php if ( Sensei_Quiz::is_reset_allowed( Sensei()->quiz->get_lesson_id( get_the_ID() ) ) ) : ?>
+			<div class="sensei-quiz-pagination__button-link">
+				<button type="submit" name="quiz_reset" class="sensei-stop-double-submission has-background has-text-color">
```
I'm not too clear if `has-background` and `has-text-color` are needed here. I think they pull in the theme's colors or something like that? Would you mind clarifying?

m1r0 replied:
The initial issue here was that some of the `twenty` themes have very specific CSS selectors for the button background color that can't be easily overwritten.

By using these classes I've managed to avoid the use of `!important` or very long selectors to overwrite the styles. 

But thinking about it now, adding theme-specific classes doesn't make much sense. I've removed the classes and changed to code to use `!important` in 8ae06031cb8b0b4e1883203ad0e9a8638ff96081. Let me know if there is a better approach.

donnapep replied:
I tested this with Twenty Twenty One. The problem appears to be that the theme's CSS selector (`button:not(:hover):not(:active):not(.has-background)`) has much higher specificity than our selector because of all the pseudo classes. Probably the easiest way to combat this would be to give the button an ID like `sensei-quiz-save` and then use that to style it.

m1r0 replied:
@donnapep Yeah, that's an option for the reset/save buttons, but we will have the same problem down the road (#4521) when the pagination links are transformed into buttons. Should I add the IDs now and figure this out later?

m1r0 replied:
I think this update https://github.com/Automattic/sensei/commit/70caded9fbc035026d2ce092cfae6d075e28c489 solves all of the above.

donnapep commented:
I noticed that the question numbers reset back to 1 as you move through the pages. I'm thinking it shouldn't do that since if I have 10 questions on a single page, they will be numbered 1 to 10.

donnapep commented:
> Make the quiz pagination per page number bigger than the questions count and make sure the pagination UI is not loading. You should see the default action buttons instead.

I'm wondering about this bit, because the views look quite a bit different depending on whether pagination is enabled or not. In particular, I'm wondering if the actions should be styled similarly in either case.

For example, here are the buttons when pagination is disabled (on Astra):

![Screen Shot 2021-12-09 at 12 01 09 PM](https://user-images.githubusercontent.com/1190420/145441950-3340bdf5-e1ac-4dd6-8c2b-1e6335217bef.jpg)

And here are the buttons when pagination is enabled:

![Screen Shot 2021-12-09 at 12 02 54 PM](https://user-images.githubusercontent.com/1190420/145442204-7dfe7aeb-b9fb-4852-8fb4-d5657ea41647.jpg)

@sruj09 Should these buttons / links be styled and named the same in both cases?

m1r0 commented:
> I noticed that the question numbers reset back to 1 as you move through the pages. I'm thinking it shouldn't do that since if I have 10 questions on a single page, they will be numbered 1 to 10.

Thanks, I've missed that one.  Fixed in 3a2a29a271a6e99bda849dda19af66069965ad06.

m1r0 commented:
> Looks like there's an issue on Divi

Fixed in 5ae2c5d94a2b3adaf007ff1e6928419207668bcb.

m1r0 commented:
> One minor update, can we add the arrow icons in the buttons instead of the < and >?

@sruj09 I've totally missed that, thanks! I've added the icons with b19f38600417cc9c5dc1b94303ddc71ec76df314 and updated the PR screenshots. I'm using the already present font awesome icons, so they might look slightly different.

sruj09 commented:
Looks good @m1r0 🙌

donnapep commented on assets/css/frontend.scss:
```diff
@@ -370,6 +370,111 @@ a.sensei-certificate-link {
+	gap: 1.5em;
+	margin: 1.618em 0;
+
+	@media only screen and (min-width: 768px) {
```
Maybe we could add a variable for 768px and use that throughout instead?

m1r0 replied:
Done in https://github.com/Automattic/sensei/pull/4502/commits/8c0e4cadd7980e8d60b3b8d64a7d893c8e2ed9cc.

donnapep commented:
I think we should go ahead and make the button / links names and styles consistent as per https://github.com/Automattic/sensei/pull/4502#issuecomment-990043429. It's a bit odd to have a different look depending on whether pagination is enabled or not.

Something else I noticed while testing is that pagination only works if I'm enrolled in the course. If I'm logged out, or logged in but not enrolled, it shows all questions at once.

donnapep commented on templates/single-quiz.php:
```diff
@@ -57,7 +57,10 @@
-					<li class="<?php sensei_the_question_class(); ?>">
+					<li
+						class="<?php sensei_the_question_class(); ?>"
+						value="<?php echo esc_attr( sensei_get_the_question_number() ); ?>"
```
Wow, I've never seen this attribute used on a list item before. 🆒 

Since we've changed a template file and some people override them with their own versions, we'll need to bump the template version, add the "Templates" label and make note of which template was changed in the PR description.

m1r0 replied:
> Wow, I've never seen this attribute used on a list item before.

Same... 😄 

>Since we've changed a template file and some people override them with their own versions, we'll need to bump the template version, add the "Templates" label and make note of which template was changed in the PR description.

Version bumped in d50a5e259f3da2446db04581920d555c6d28bf53, thanks!

m1r0 commented:
> Something else I noticed while testing is that pagination only works if I'm enrolled in the course. If I'm logged out, or logged in but not enrolled, it shows all questions at once.

@donnapep Changed it so the pagination is enabled for users that are not enrolled (aed041321b1f9a3dd73031bc447e8d9e1bfb91d9). If you are logged out, I think the default behavior is to not show any questions if the lesson is not public.

m1r0 commented:
> I think we should go ahead and make the button / links names and styles consistent.

@donnapep I've made a [new issue](https://github.com/Automattic/sensei/issues/4540) for this, so we can keep this PR a bit smaller.

donnapep commented:
> Should I use a template variable or just make the `can_take_quiz()` method public and call it from the template file?

Hmmm. Yeah, good question. I think I'd slightly lean towards making the function public in this case, just because it does seem like it would be a useful function for other developers to be able to call.

m1r0 commented:
> Hmmm. Yeah, good question. I think I'd slightly lean towards making the function public in this case, just because it does seem like it would be a useful function for other developers to be able to call.

@donnapep Yeah makes sense. I've applied the change in 0f16b609bc12a5bd35767b4908cabed05ced26cd.

donnapep reviewed (approved):
🍾 🎆
